### PR TITLE
Refactor sky rotation classes to use the EulerAngleRotation class.

### PR DESCRIPTION
**Pull request by nden**
The sky rotation classes were implemented to as subclasses of `EulerAngleRotation`. This allows to avoid dealing with special cases and increases precision.
In response to spacetelescope/gwcs#50.

Not sure who's best to review or whom to assign it.
@philhodge @mcara @eteq

### CODE REVIEW

**Comment by nden** [comment]
Since this is an internal change I'm not sure it needs an entry in the change log.

**Comment by astrofrog** [comment]
@nden - it looks like the test failure is real

**Comment by mcara** [comment]
> Not sure who's best to review or whom to assign it. @philhodge @mcara @eteq

It seems (see https://github.com/spacetelescope/gwcs/issues/50#issuecomment-218446250) that @philhodge got a deep understanding of matrix representation of Euler rotations.

**Comment by philhodge** [comment]
On 05/13/2016 10:21 AM, Mihai Cara wrote:
> Not sure who's best to review or whom to assign it. @philhodge 
> <https://github.com/philhodge> @mcara <https://github.com/mcara> @eteq 
> <https://github.com/eteq>

I'll look at it.

Phil

**Comment by mcara** [comment]
I was also wondering if a different test may be more "interesting" such as (using notations from the code of https://github.com/spacetelescope/gwcs/issues/50#issue-153580789):

```python
>>> # generate test RA & DEC using three categories of values:
>>> # 1. randomly distributed across the sphere
>>> # 2. points exactly at the pole (CRVAL) and at/near the opposite pole
>>> # 3. randomly distributed points near (within 1 pixel) from the poles
>>> #    (a pixel could be taken to be 1.6666e-5 arcdeg in scale)
>>>
>>> test_ra, test_dec = get_test_points(fwi)  # pass fwi to get pole location, pixel scale, etc.
>>> xy_stdwcs = fwi.wcs_world2pix([test_ra, test_dec], 1)
>>> xy_gwcs = np.asanyarray(gwi.invert(test_ra, test_dec)).T
>>> utils.assert_allclose(xy_stdwcs, xy_gwcs, rtol=1e-10)
```

Something like that... This test should not replace the round-trip test but be *in addition* to it, at least for the time being.

**Comment by nden** [comment]
@philhodge I'm working on some changes.

**Comment by nden** [comment]
@philhodge Travis has a queue, I'm not sure when it will get to this PR but tests pass locally for me. This is ready for review now.

@mcara There's a test comparing with `wcslib`.

File: astropy/modeling/rotations.py [hunk 1]
```diff
@@ -183,39 +167,56 @@ class RotateNative2Celestial(_SkyRotation):
...
+        lon = -(self.psi + 90)
+        lat = self.theta + 90
+        lon_pole = self.phi + 90
+        return RotateCelestial2Native(lon, lat, lon_pole)
 
 
 class RotateCelestial2Native(_SkyRotation):
     """
-    Transform from Celestial to Native to Spherical Coordinates.
+    Transform from Native to Celestial Spherical Coordinates.
```
Comment (philhodge) [inline astropy/modeling/rotations.py]: "Native" and "Celestial" appear to be swapped in this comment.

**Comment by eteq** [comment]
I gave this a quick look-over and it seems ok to me, @nden.

 One question: does this need a changelog entry? I think it's nominally a change to the public API because it changes the class heirarchy of a public class, right?  It probably won't really matter much to almost anyway, but in the name of completeness I think that's still formally-speaking an API change.

**Comment by nden** [comment]
AppVeyor tests failed because `scipy` cannot be imported. Assuming it's a glitch, can someone restart those tests?

**Pull request merged**